slf4j LogAppender

justcoon opened the pull request:
slf4j log appender 

- key-values - `LogAppender.appendKeyValue` - logged into MDC
- cause - `LogAppender.appendCause` - as throwable, `Cause` transformed into `FiberFailure`
- slf4j logger name extracted from `Trace`

open question -
- do we need `rootLogLevel` in `private def slf4jLogger(rootLoggerName: Trace => String, rootLogLevel: LogLevel, format: LogFormat): ZLogger[String, Unit]` ? - in general, log level setup is in specific logger configuration
- `LogFormat` combinators/composition - in current design, it is not clear, where specific LogFormat part will be "inserted" (message, MDC, cause ...), there is a question, if it should not be more clear in LogFormat itself
- should we allow custom definition for log logger name (maybe by annotation as it was in version 1) ? 

related to: #485, #453, #445

987Nabil commented:
@justcoon We impl. our service based on the `rootLogLevel` in private def `slf4jLogger(rootLoggerName: Trace => String, rootLogLevel: LogLevel, format: LogFormat): ZLogger[String, Unit]`. This was the way I found to make different logger settings based on the logger name work.

For example
```xml
<logger name="access" level="INFO" additivity="false">
  <appender-ref ref="ACCESS_CONSOLE_LOGGER"/>
</logger>
```

And the default logic you impl. is not useful
```scala
  def getLoggerName(default: String = "zio-slf4j-logger"): Trace => String =
    _ match {
      case Trace(location, _, _) =>
        val last = location.lastIndexOf(".")
        if (last > 0) {
          location.substring(0, last)
        } else location

      case _ => default
    }
```

Is there another way to do this? If not, I'd like to have this working (soon) 🙂

987Nabil commented:
@justcoon I guess this is also the reason, why updating to 2.0.1 broke our slf4j + logstash json logging?
Any plan when this would be fixed? Or do you need some help?

And to answer your last question: Yes we definitely need the ability to set logger names. In the 2xRC version we did this with the `defaultLoggerName: Trace => String`

justcoon commented:
hello @987Nabil, 

in relation to 

> 2.0.1 broke our slf4j + logstash json logging

do you mean update from 2.0.0 -> 2.0.1? If yes, maybe it will be good if you open some ticket with details (like logstash configuration, zio logging slf4j setup - log format ..., code snippet, if possible)

in case of logger name - `SLF4J.getLoggerName`
basically if there is trace `zio.logging.example.LivePingService.ping(PingService.scala:22)`, then logger name is `zio.logging.example.LivePingService` - service class name

in my opinion, this is reasonable default logger name resolution, based on trace
but at the other side, i think, have option to say, what is logger name, can be very usefull

jdegoes commented on core/shared/src/main/scala/zio/logging/internal/LogAppender.scala:
```diff
@@ -46,16 +46,11 @@ private[logging] trait LogAppender { self =>
-   * Appends a key/value pair, with the value it created with the log appender.
-   */
-  final def appendKeyValue(key: String, appendValue: LogAppender => Unit): Unit = {
+  def appendKeyValue(key: String, value: String): Unit = {
```
Let's not modify the `appendKeyValue`. Instead, generate an anonymous `LogAppender` that will simply produce string for any possible combination of operations.

justcoon replied:
@jdegoes: previously removed function def appendKeyValue(key: String, appendValue: LogAppender => Unit) of LogAppender- returned back

987Nabil commented:
> hello @987Nabil,
> 
> in relation to
> 
> > 2.0.1 broke our slf4j + logstash json logging
> 
> do you mean update from 2.0.0 -> 2.0.1? If yes, maybe it will be good if you open some ticket with details (like logstash configuration, zio logging slf4j setup - log format ..., code snippet, if possible)
I'll do.
> in case of logger name - `SLF4J.getLoggerName` basically if there is trace `zio.logging.example.LivePingService.ping(PingService.scala:22)`, then logger name is `zio.logging.example.LivePingService` - service class name
> 
> in my opinion, this is reasonable default logger name resolution, based on trace but at the other side, i think, have option to say, what is logger name, can be very usefull
Maybe it is a reasonable default, but in my opinion having it configurable is a must. There are architecture guidelines out there in the wild, that require special logger names for access logging for example. One reason for something like this can be, that messages with a certain name get treated in your logging tool based on logger names.
Also it is available in every (JVM) logging lib out there and I do not see any reason to break this pattern.

987Nabil commented:
@justcoon I opened #489 and would not be surprised, if it is related.

justcoon commented:
latest changes
* previously removed function  `def appendKeyValue(key: String, appendValue: LogAppender => Unit)` of `LogAppender`- returned back
* added `SLF4J.loggerName` aspect, for custom slf4j logger name
* layers, with `LogLevel` are deprecated 
* there are scala-2 tests only, for slf4j, as there is issue with `Trace` in scala-3 (https://github.com/zio/zio/pull/7177)
* updated documentation and examples
